---
title: "Functional analysis"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
params:
  voi : ""
---

Functional analysis is performed using PICRUSt. PICRUSt is designed to estimate the gene families contributed to a metagenome by bacteria or archaea identified using 16S rRNA sequencing. PICRUSt uses an extended ancestral-state reconstruction algorithm to predict which gene families are present and then combines gene families to estimate the composite metagenome. Two categories are investigated: KEGG and COG. 

# KEGG 

KEGG is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies.

## Barplots

### Aggregated

<div>
<object data="Figures/Taxa_DifferentialAbundance_AbundanceBarplot_none_fdr_0.1_KEGG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>


### Individual

<div>
<object data="Figures/Taxa_Barplot_All_P_fdr_0.1_KEGG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Boxplots 

### Aggregated

<div>
<object data="Figures/Taxa_DifferentialAbundance_AbundanceBoxplot_none_fdr_0.1_KEGG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

### Individual

<div>
<object data="Figures/Taxa_Boxplot_All_P_fdr_0.1_KEGG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Effect size plot

<div>
<object data="Figures/Taxa_DifferentialAbundance_logPBarplot_fdr_0.1_KEGG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Test results

```{r, echo=FALSE}
mytab <- read.csv("Figures/Taxa_DifferentialAbundanceAnalysis_AllLevels_fdr_0.1_KEGG_perm.csv", head=TRUE, sep=",")
names(mytab)[1] <- "Taxa"
knitr::kable(mytab[,c(1,2,3,4,5,8)], caption="Differential abundance analysis for all levels")
```

# COG

The COG protein database was generated by comparing predicted and known proteins in all completely sequenced microbial genomes to infer sets of orthologs. Each COG consists of a group of proteins found to be orthologous across at least three lineages and likely corresponds to an ancient conserved domain.

## Barplots

### Aggregated

<div>
<object data="Figures/Taxa_DifferentialAbundance_AbundanceBarplot_none_fdr_0.1_COG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

### Individual

<div>
<object data="Figures/Taxa_Barplot_All_P_fdr_0.1_COG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Boxplots 

### Aggregated

<div>
<object data="Figures/Taxa_DifferentialAbundance_AbundanceBoxplot_none_fdr_0.1_COG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

### Individual

<div>
<object data="Figures/Taxa_Boxplot_All_P_fdr_0.1_COG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Effect size plot

<div>
<object data="Figures/Taxa_DifferentialAbundance_logPBarplot_fdr_0.1_COG_.pdf" type="application/pdf" width="900" height="600">
</object>
</div>

## Test results

```{r, echo=FALSE}
mytab <- read.csv("Figures/Taxa_DifferentialAbundanceAnalysis_AllLevels_fdr_0.1_COG_perm.csv", head=TRUE, sep=",")
names(mytab)[1] <- "Taxa"
knitr::kable(mytab[,c(1,2,3,4,5,8)], caption="Differential abundance analysis for all levels")
```

